File: Kern/Sources/DuoKern/Phrase/PhraseTemplate.swift
/// Slot-template model for generated phrases (design.md Phase 3 tail):
/// a curated sentence frame whose single `{slot}` is filled with a
/// Trainer-generated value — digits on the German prompt side,
/// target-language words on the answer side.
public struct PhraseTemplate: Sendable, Equatable, Identifiable {
    /// Ukrainian counted-noun agreement for a `{count}` marker following the
    /// numeral: 1/21/31… → `one`, 2–4/22–24… → `few`, everything else
    /// (incl. 11–14) → `many`.
    public struct CountForms: Sendable, Equatable {
        public var one: String
        public var few: String
        public var many: String

        public init(one: String, few: String, many: String) {
            self.one = one
            self.few = few
            self.many = many
        }

        public func form(for n: Int) -> String {
            let lastTwo = abs(n) % 100
            if (11...14).contains(lastTwo) { return many }
            switch lastTwo % 10 {
            case 1: return one
            case 2, 3, 4: return few
            default: return many
            }
        }
    }

    public var id: String
    public var pair: LanguagePair
    /// German sentence with `{slot}`; the prompt substitutes digits
    /// ("14:35", "347", "1978"), never words.
    public var deTemplate: String
    /// Target sentence with `{slot}` (and `{count}` iff `countForms` is set);
    /// display/accepted substitute the Trainer's word forms.
    public var targetTemplate: String
    public var slotKind: TrainerKind
    /// Template-level note, merged with the slot task's gloss.
    public var gloss: String?
    /// Present only on `.numbers` templates whose target noun must agree
    /// with the numeral (Ukrainian). `targetTemplate` then contains `{count}`.
    public var countForms: CountForms?
    /// Ukrainian templates counting a masculine/indeclinable noun: feminine
    /// numeral variants (одна/дві) must NOT be accepted — these templates
    /// exist to train exactly that agreement (language-review finding).
    /// Implied for all `countForms` templates.
    public var masculineSlot: Bool

    public init(id: String, pair: LanguagePair, deTemplate: String, targetTemplate: String,
                slotKind: TrainerKind, gloss: String? = nil, countForms: CountForms? = nil,
                masculineSlot: Bool = false) {
        self.id = id
        self.pair = pair
        self.deTemplate = deTemplate
        self.targetTemplate = targetTemplate
        self.slotKind = slotKind
        self.gloss = gloss
        self.countForms = countForms
        self.masculineSlot = masculineSlot || countForms != nil
    }

    public var targetLanguage: TrainerLanguage {
        switch pair {
        case .deSw: return .swahili
        case .deUk: return .ukrainian
        }
    }

    static let slotMarker = "{slot}"
    static let countMarker = "{count}"
}

File: kern/src/commonMain/kotlin/net/spross/kern/trainer/PhraseTemplate.kt
package net.spross.kern.trainer

import net.spross.kern.model.Language

/**
 * Slot-template model for generated phrases: a curated sentence frame whose
 * single `{slot}` is filled with a Trainer-generated value — digits on the
 * source-language prompt side, target-language words on the answer side.
 */
data class PhraseTemplate(
    val id: String,
    /** Known-language side of the frame (the prompt on forward drills). */
    val source: Language,
    /** Learning-language side; also selects the slot generator. */
    val target: Language,
    /**
     * Source sentence with `{slot}` (and `{count}` iff [sourceCountForms] is set);
     * the prompt substitutes digits ("14:35", "347", "1978"), never words.
     */
    val sourceTemplate: String,
    /**
     * Target sentence with `{slot}` (and `{count}` iff [countForms] is set);
     * display/accepted substitute the Trainer's word forms.
     */
    val targetTemplate: String,
    val slotKind: TrainerKind,
    /**
     * Accept-only alternate renderings of [targetTemplate] (the du/Sie register split):
     * graded as correct, never displayed.
     */
    val acceptedFrames: List<String> = emptyList(),
    /**
     * The answer realization's note in the learner's explanation language,
     * merged with the slot task's gloss.
     */
    val note: String? = null,
    /**
     * Present only on Numbers templates whose target noun must agree with the
     * numeral (Ukrainian). [targetTemplate] then contains `{count}`.
     */
    val countForms: CountForms? = null,
    /**
     * [countForms] on the source side. Agreement is a property of the realization, so a
     * frame authored with `{count}` carries it into every pair — including the ones where
     * that language is the PROMPT and the marker would otherwise reach the learner literally.
     */
    val sourceCountForms: CountForms? = null,
    /**
     * Templates counting a masculine/indeclinable noun: feminine numeral
     * variants (одна/дві) must NOT be accepted — these templates exist to
     * train exactly that agreement (language-review finding).
     */
    val masculineNumeral: Boolean = false,
    /**
     * Swahili templates counting a noun whose class the numeral agrees with: the Bantu
     * cardinals render with that class's concord prefix ([SwahiliConcord]).
     * Null everywhere else — including the Swahili N-class frames already shipping
     * (*sahani*, *funguo*), whose numerals are the bare ones [SwahiliNumbers] spells.
     */
    val swahiliNounClass: SwahiliConcord.NounClass? = null,
) {
    init {
        // Forms readings are not sentence slots: a fraction or ordinal needs the frame to
        // decline around it, and no agreement device runs that way (docs/backlog.md).
        // Failing here means catalog build time, not the draw that would have thrown.
        require(slotKind != TrainerKind.Forms) { "$id: frames cannot carry a ${TrainerKind.Forms} slot" }
        // Noun-class concord is Swahili's and needs a numeral to carry it — a frame that
        // claims it anywhere else would silently render the plain reading instead.
        require(swahiliNounClass == null || (slotKind == TrainerKind.Numbers && target == "sw")) {
            "$id: swahiliNounClass needs a Swahili ${TrainerKind.Numbers} slot, not $target/$slotKind"
        }
    }

    /** Effective masculine-numeral rule — implied for all [countForms] templates. */
    val masculineNumeralOnly: Boolean get() = masculineNumeral || countForms != null

    /**
     * Ukrainian counted-noun agreement for a `{count}` marker following the
     * numeral: 1/21/31… → [one], 2–4/22–24… → [few], everything else
     * (incl. 11–14) → [many].
     */
    data class CountForms(val one: String, val few: String, val many: String) {
        fun form(n: Long): String {
            val lastTwo = (if (n < 0) -n else n) % 100
            if (lastTwo in 11..14) return many
            return when (lastTwo % 10) {
                1L -> one
                2L, 3L, 4L -> few
                else -> many
            }
        }
    }

    companion object {
        const val SLOT_MARKER = "{slot}"
        const val COUNT_MARKER = "{count}"
    }
}
